vars/withTaskQueue.groovy

```diff
@@ -0,0 +1,149 @@
+  withTaskQueue(parallel: 10, setup: { sh "my-setup-scrupt.sh" }) {
+    ...
+  }
+
+*/
+def call(Map options = [:], Closure closure) {
+  def config = [ parallel: 10 ] + options
+  def counter = ++TASK_QUEUES_COUNTER
+
+  // We're basically abusing withEnv() to create a "scope" for all steps inside of a withTaskQueue block
+  // This way, we could have multiple task queue instances in the same pipeline
+  withEnv(["TASK_QUEUE_ID=${counter}"]) {
+    withTaskQueue.TASK_QUEUES[env.TASK_QUEUE_ID] = [
+      tasks: [],
+      tmpFile: sh(script: 'mktemp', returnStdout: true).trim()
+    ]
+
+    closure.call()
+
+    def processesExecuting = 0
+    def processes = [:]
+    def iterationId = 0
+
+    for(def i = 1; i <= config.parallel; i++) {
+      def j = i
+      processes["task-queue-process-${j}"] = {
+        catchErrors {
+          withEnv([
+            "TASK_QUEUE_PROCESS_ID=${j}",
+            "TASK_QUEUE_ITERATION_ID=${++iterationId}"
+          ]) {
+            dir("${WORKSPACE}/parallel/${j}/kibana") {
+              if (config.setup) {
+                config.setup.call(j)
+              }
+
+              def isDone = false
+              while(!isDone) { // TODO some kind of timeout?
+                catchErrors {
+                  if (!getTasks().isEmpty()) {
+                    processesExecuting++
+                    catchErrors {
+                      def task
+                      try {
+                        task = getTasks().pop()
+                      } catch (java.util.NoSuchElementException ex) {
+                        return
+                      }
+
+                      task.call()
+                    }
+                    processesExecuting--
+                    // If a task finishes, and no new tasks were queued up, and nothing else is executing
+                    // Then all of the processes should wake up and exit
+                    if (processesExecuting < 1 && getTasks().isEmpty()) {
+                      taskNotify()
+                    }
+                    return
+                  }
+
+                  if (processesExecuting > 0) {
+                    taskSleep()
+                    return
+                  }
+
+                  // Queue is empty, no processes are executing
+                  isDone = true
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    parallel(processes)
+  }
+}
+
+// If we sleep in a loop using Groovy code, Pipeline Steps is flooded with Sleep steps
+// So, instead, we just watch a file and `touch` it whenever something happens that could modify the queue
+def taskSleep() {
+  sh(script: """#!/bin/bash
+    TIMESTAMP=\$(date '+%s' -d "0 seconds ago")
+    for (( i=1; i<=240; i++ ))
+    do
+      if [ "\$(stat -c %Y '${getTmpFile()}')" -ge "\$TIMESTAMP" ]
+      then
+        break
+      else
+        sleep 5
+      fi
+    done
```

Comment:
Do you think there's a chance this could indicate to the outside world when it's waited for 20 minutes and never found changes to the tmp file? Should that be an error condition? It feels like `taskSleep()` should either wait until the task is needed or timeout, but maybe it just needs a comment to explain why there's a 20 minute timeout that doesn't fail when it's triggered.

Reply:
I can add a message. It's just a safety to keep it from infinite looping in case something goes wrong. The outer groovy loop will actually call this sleep script again if the 20-minute timeout triggers and it still actually is supposed to be waiting.

Reply:
Gotcha, yeah, seems weird to me that it would end at all. Seems like an opportunity to miss the timestamp update from `taskNotify()`. If you used something like `while true; do` wouldn't Jenkins kill the executor if the overall timeout was hit?